Server/PlayerLogic.py

```diff
@@ -2,58 +2,151 @@
-            self._send_updated_pos()
-        if self.initial_pos_sending:
-            self._send_updated_pos()
-            self.initial_pos_sending = False
+    def _add_force(self, force2d):
+        self._forces.append(force2d)
 
-    def get_id(self):
-        return self._id
+    def _add_ground_directed_force(self, force_mag, direction):
+        ang = self._terrain.get_angle_rad(self._pos.x)
+        force = Vector2D.mag_ang_init(force_mag*direction.value, ang)
+        self._add_force(force)
+
+    def my_dir(self):
+        if self._vel.x > 0:
+            return Direction.RIGHT
+        return Direction.LEFT
+
+    def can_accelerate(self, direction=1):
+        if direction*self._vel.x < 0:  # it can always decelerate
+            return True
+        base_max_vel = 3.0
+        angle_dependency = 3.0
+        angle_bonus = -direction * math.sin(self._terrain.get_angle_rad(self._pos.x)) * angle_dependency
+        return self._vel.mag() < base_max_vel + angle_bonus
+
+    def _move_left(self):
+        if self.can_accelerate(-1):
+            self._add_ground_directed_force(self._mobility, Direction.LEFT)
+
+    def _move_right(self):
+        if self.can_accelerate(1):
+            self._add_ground_directed_force(self._mobility, Direction.RIGHT)
+
+    def _stop(self):
+        self._vel = Vector2D.zero()
+
+    def _put_to_ground_level(self):
+        self._pos.y = self._terrain.get_level(self._pos.x)+self.RADIUS
+
+    def _get_y_to_ground_level(self):
+        return self._pos.y - (self._terrain.get_level(self._pos.x)+self.RADIUS)
+
+    def _impact(self):
+        pass  # BUMM!
+
+    def _do_physics(self):
+        # Add some environmental forces
+        if self._is_flying:
+            # add gravitation
+            mg = self.G * self._mass
+            self._add_force(Vector2D(0, mg))
+        else:
+            # add friction
+            friction = self._mass * self.mu  # yes it's not perfectly accurate physics
+            friction = min(self._vel.mag(), friction)  # friction shouldn't accelerate to the opposite direction
+            self._add_ground_directed_force(-friction, self.my_dir())
+
+        # Compute main properties!
+        resultant_force = Vector2D.sum(self._forces)
+        self._forces = []
+        res_acc = resultant_force.scalar_div(self._mass)
+        self._vel += res_acc
+        self._pos += self._vel
+
+        # Check if it flies
+        threshold_ang = 2*3.14/180.0  # 2 degrees
+        threshold_pos = 2  # 2 pixels
+        threshold_big_pos = 5  # 5 pixels alone is surely flying
+        vel_angle_diff = math.fabs(self._vel.domi_ang()-self._terrain.get_angle_rad(self._pos.x))
+        pos_diff = self._get_y_to_ground_level()
+        if (vel_angle_diff > threshold_ang and pos_diff > threshold_pos)\
+                or pos_diff > threshold_big_pos:
+            if not self._is_flying:
+                self._add_force(Vector2D(0, 10))  # for fun :)
+            self._is_flying = True
```

Comment:
Again, _threshold_ang, threshold_pos, threshold_big_poss_ should be defined as constants, and a more straightforward description of their use in a comment after them would be appreciated. :) 
Also, instead of _# for fun :)_ - a description of what is intended to happen would be welcome :D

Reply:
:D